### PR TITLE
[lazy-defs] Add reconstruction metadata to PowerBI integration

### PR DESCRIPTION
**Pull request by smackesey**
## Summary & Motivation

Add a reconstruction metadata API to the PowerBI integration.

Also make the existing PowerBI data structures serializable to eliminate an extra layer of translation to plain dict/list for serialization.

The PowerBI integration already offers state-driven asset definitions using `CacheableAssetsDefinition`. This adds an alternative `load_powerbi_defs` function as an alternative that uses code location instead of `CacheableAssetsDefinition`. It is invoked like this:

```python
@definitions
def reconstruction_metadata_defs(context: DefinitionsLoadContext) -> Definitions:
    pbi_defs = load_powerbi_defs(
        context,
        api_token="token",
        workspace_id="my_workspace",
        enable_refresh_semantic_models=False,
    )

    return Definitions.merge(
        Definitions(assets=[my_materializable_asset], jobs=[define_asset_job("all_asset_job")]),
        pbi_defs,
    )
```

One outstanding question is how exactly to structure the API for `load_powerbi_defs`. I opted for a simple exported function that doesn't make you deal with any other abstractions, but the existing `CacheableAssetsDefinition` implementation has you do this:

```
cacheable_asset_defs = PowerBIWorkspace(
    api_token=...,
    workspace_id=...,
).build_defs()
```

It would be easy to use a similar API for the code location metadata version. I'm agnostic on this and we should do whatever is more idiomatic for our integrations.

## How I Tested These Changes

New unit test that follows same template as the test on `CacheableAssetsDefinition`.

## Changelog

- [ ] `NEW` _(added new feature or capability)_
- [ ] `BUGFIX` _(fixed a bug)_
- [ ] `DOCS` _(added or updated documentation)_

### CODE REVIEW

**Comment by smackesey** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24403?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#24403** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24403?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#24395** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24395?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24397** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24397?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#24337** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24337?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @smackesey and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: python_modules/libraries/dagster-powerbi/dagster_powerbi_tests/pending_repo.py [hunk 1]
```diff
@@ -28,3 +31,18 @@ def my_materializable_asset():
...
+        context,
+        api_token=fake_token,
+        workspace_id="a2122b8f-d7e1-42e8-be2b-a5e636ca3221",
+        enable_refresh_semantic_models=False,
+    )
+
+    return Definitions.merge(
+        Definitions(assets=[my_materializable_asset], jobs=[define_asset_job("all_asset_job")]),
+        pbi_defs,
+    )
```
Comment (schrockn) [inline python_modules/libraries/dagster-powerbi/dagster_powerbi_tests/pending_repo.py:48]: <i class='graphite__hidden'>[Re: lines +36 to +49]<br/><br/></i>

Much better<span class='graphite__hidden'><br/><br/>See this comment inline on <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/24403?utm_source=unchanged-line-comment">Graphite</a>.</span>

**Review by schrockn — CHANGES_REQUESTED** [review]
restack on other proposed changes

File: python_modules/libraries/dagster-powerbi/dagster_powerbi/resource.py [hunk 2]
```diff
@@ -233,6 +234,68 @@ def build_defs(
...
+) -> Sequence[Mapping[str, Any]]:
+    return [
+        data.to_cached_data()
+        for data in [
+            *workspace_data.dashboards_by_id.values(),
+            *workspace_data.reports_by_id.values(),
+            *workspace_data.semantic_models_by_id.values(),
+            *workspace_data.data_sources_by_id.values(),
+        ]
+    ]
```
Comment (schrockn) [inline python_modules/libraries/dagster-powerbi/dagster_powerbi/resource.py:248]: Lets make this typed `record`. We should do that as a norm so we can evolve and document the serialization formats.

File: python_modules/libraries/dagster-powerbi/dagster_powerbi/resource.py [hunk 3]
```diff
@@ -233,6 +234,68 @@ def build_defs(
...
+            power_bi.trigger_refresh(dataset_id)
+            power_bi.poll_refresh(dataset_id)
+            context.log.info("Refresh completed.")
+
+        executable_assets.append(asset_fn)
+
+    return [*external_assets_from_specs(all_external_asset_specs), *executable_assets]
+
+
 class PowerBICacheableAssetsDefinition(CacheableAssetsDefinition):
```
Comment (schrockn) [inline python_modules/libraries/dagster-powerbi/dagster_powerbi/resource.py:299]: can't we just kill this?

**Review by schrockn — CHANGES_REQUESTED** [review]
ineline comments

File: python_modules/dagster/dagster/_core/definitions/cacheable_assets.py [hunk 4]
```diff
@@ -63,9 +64,9 @@ def __new__(
     ):
         extra_metadata = check.opt_nullable_mapping_param(extra_metadata, "extra_metadata")
         try:
-            # check that the value is JSON serializable
-            seven.dumps(extra_metadata)
-        except TypeError:
+            # check that the value can pass through the serdes layer
+            serialize_value(extra_metadata)
+        except SerializationError:
```
Comment (smackesey) [inline python_modules/dagster/dagster/_core/definitions/cacheable_assets.py:69]: PowerBI objects were failing serialization without this change-- `@record` objects aren't iterable and so `json.dumps` chokes on them, they need to undergo the preprocessing in `serialize_value`.

File: python_modules/libraries/dagster-powerbi/dagster_powerbi/resource.py (hunk 3 above)
Comment (smackesey) [reply python_modules/libraries/dagster-powerbi/dagster_powerbi/resource.py:285]: as discussed offline not rming in this pr

**Review by schrockn — APPROVED** [review]
Great step forward. I would like @benpankow to take a pass before merging.

**Comment by schrockn** [comment]
Took a stab at the slightly higher level pattern to structure the interaction with the context a little more here: https://github.com/dagster-io/dagster/pull/24496

**Review by benpankow — APPROVED** [review]
This looks good to me. I would prefer to structure this as a method on the resource, something like `build_lazy_defs(...)` (e.g. for Sigma we just define `build_defs` using this new pattern: https://github.com/dagster-io/dagster/pull/24516) - I am happy handle this in a follow-up though.

**Comment by smackesey** [comment]
>I am happy handle this in a follow-up though.

Cool, gonna merge then!

**Pull request merged**